flamegraph/flamegraph.py

```diff
@@ -81,7 +83,6 @@ def _write_results(self):
       self._written = True
       for key in sorted(self._stats.keys()):
         self._fd.write('%s %d\n' % (key, self._stats[key]))
-      self._fd.close()
```

Comment:
Why this deletion?

Reply:
The tests that rely on `capsys` cannot pass otherwise: closing the stream prevents reading from it near the end of test.

I'm actually curious to know why you put this in the first place since standard output is closed automatically :-)

Reply:
@AndreLouisCaron This doesn't look correct. If the user invokes flamegraph like `python -m flamegraph -o profile.log`, then `self._fd` will be a file, which won't be closed automatically.

Reply:
Aren't all file descriptors are closed automatically (by the OS) when the process exits?

Reply:
Yes, but this doesn't necessarily mean that all buffers are flushed.

Python buffers writes to files. In the following test program, calling the write() method on the file does not result in a write() syscall until the file is closed.

```python
import time
f = open("foo", "wt")
f.write("test")
time.sleep(10)
f.close()
```

You can test this with

```bash
$ strace -e write,open,close,openat python3 test.py
```

The write() syscall happens after the ten-second wait, despite the write() method happening before the sleep.

In CPython, a close() will happen eventually no matter what. (Even if you comment out `f.close()`, the close syscall will still happen.) However, I think that's an implementation-defined behavior, and not something you should rely on.

See also [Is explictly closing files important?](https://stackoverflow.com/questions/7395542/is-explicitly-closing-files-important), [Python docs](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files)